File: Study Plan/Top Interview 150/0242 Valid Angram/ValidAngram.java
/*  
    Problem
    --------------------------------------------------------------------------------
    242. Valid Angram
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/valid-anagram/solutions/4529354/java-beats-97-in-runtime-time-complexity-o-n-log-n-dual-pivot-quicksort/

    - Runtime 2 ms (Beats 97.03% of users with Java)
    - Memory 44.52 MB (Beats 19.23% of users with Java)
*/

class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        char[] source = s.toCharArray();
        char[] target = t.toCharArray();
        /* 
         *  The sorting algorithm is a Dual-Pivot Quicksort.
         *  This algorithm offers O(n log(n)) performance on all data sets,
         *  and is typically faster than traditional (one-pivot) Quicksort
         *  implementations.
         *
         *  https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/DualPivotQuicksort.java
         */
        Arrays.sort(source);
        Arrays.sort(target);
        return new String(source).equals(new String(target));
    }
}

File: Study Plan/Top Interview 150/0242 Valid Angram/ValidAngram.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    242. Valid Angram
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/valid-anagram/solutions/4649560/c-beats-87-in-runtime-time-complexity-o-n-log-n-introspective-sort/
    
    - Runtime 58 ms (Beats 87.36% of users with C#)
    - Memory 42.54 MB (Beats 17.00% of users with C#)
*/

public class Solution
{
    public bool IsAnagram(string s, string t)
    {
        if (s.Length != t.Length) return false;
        
        var source = s.ToCharArray();
        var target = t.ToCharArray();

        /*
            This method uses the Introspective Sort (Introsort) algorithm as
            follows:

            - If the partition size is less than or equal to 16 elements, it
            uses an Insertion Sort algorithm.

            - If the number of partitions exceeds 2 log(n), where n is the range
            of the input array, it uses a Heapsort algorithm.

            - Otherwise, it uses a Quicksort algorithm.
            This implementation performs an unstable sort; that is, if two
            elements are equal, their order might not be preserved.
            In contrast, a stable sort preserves the order of elements that are
            equal.
            This method is an O(n log(n)) operation, where n is length.

            https://learn.microsoft.com/en-us/dotnet/api/system.array.sort?view=net-8.0
        */
        Array.Sort(source);
        Array.Sort(target);

        return new string(source) == new string(target);
    }
}
